src/libs/Queue/Queue.ts

```diff
@@ -0,0 +1,68 @@
+import type Queue from './QueueType';
+
+// Function to create a new queue
+function createQueue<T>(processItem: (item: T) => Promise<void>): Queue<T> {
+    // Array to hold the elements of the queue
+    const elements: T[] = [];
+    let isProcessing = false;
+
+    // Function to remove an item from the front of the queue
+    function dequeue(): T | undefined {
+        return elements.shift();
+    }
+
+    // Function to check if the queue is empty
+    function isEmpty(): boolean {
+        return elements.length === 0;
+    }
+
+    // Initiates the processing of items in the queue.
+    // Continues to dequeue and process items as long as the queue is not empty.
+    // Sets the `isProcessing` flag to true at the start and resets it to false once all items have been processed.
+    function run(): Promise<void> {
+        return new Promise((resolve) => {
+            isProcessing = true;
+            function processNext() {
+                if (!isEmpty()) {
+                    const nextItem = dequeue();
+                    if (nextItem) {
+                        processItem(nextItem).then(processNext);
+                    }
+                } else {
+                    isProcessing = false;
+                    resolve();
+                }
+            }
```

Comment:
Any reason why you're defining this function here? I think it will get defined every time you'd call `run` function.

Reply:
it has been refactored to:

```

    // Function to process the next item in the queue
    function processNext(): Promise<void> {
        return new Promise((resolve) => {
            if (!isEmpty()) {
                const nextItem = dequeue();
                if (nextItem) {
                    processItem(nextItem).then(() => processNext().then(resolve));
                }
            } else {
                isProcessing = false;
                resolve();
            }
        });
    }

    // Initiates the processing of items in the queue.
    // Continues to dequeue and process items as long as the queue is not empty.
    // Sets the `isProcessing` flag to true at the start and resets it to false once all items have been processed.
    function run(): Promise<void> {
        isProcessing = true;
        return processNext();
    }
```